reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt

```diff
@@ -0,0 +1,256 @@
+
+        downstream.subscribe(
+            object : DefaultCompletableObserver {
+                override fun onError(error: Throwable) {
+                    isResourceDisposedBeforeInnerCompletable = resource.isDisposed
+                }
+            }
+        )
+
+        upstream.onError(Exception())
+
+        assertTrue(isResourceDisposedBeforeInnerCompletable)
+    }
+
+    @Test
+    fun disposes_resource_WHEN_eager_is_false_and_downstream_disposed() {
+        val resource = Disposable()
+        val observer = completableUsing(resourceSupplier = { resource }, eager = true).test()
+
+        observer.dispose()
+
+        assertTrue(resource.isDisposed)
+    }
+
+    @Test
+    fun disposes_resource_WHEN_eager_is_false_and_upstream_completed() {
+        val resource = Disposable()
+        val upstream = TestCompletable()
+        completableUsing(resourceSupplier = { resource }, eager = true, sourceSupplier = { upstream }).test()
+
+        upstream.onComplete()
+
+        assertTrue(resource.isDisposed)
+    }
+
+    @Test
+    fun disposes_resource_WHEN_eager_is_false_and_upstream_produced_error() {
+        val resource = Disposable()
+        val upstream = TestCompletable()
+        completableUsing(resourceSupplier = { resource }, eager = false, sourceSupplier = { upstream }).test()
+
+        upstream.onError(Exception())
+
+        assertTrue(resource.isDisposed)
+    }
+
+    @Test
+    fun resource_is_not_disposed_WHEN_eager_is_false_and_upstream_disposed() {
+        var isResourceDisposedBeforeInnerCompletable by AtomicBoolean()
+
+        val observer =
+            completableUsing(eager = false) { resource ->
+                completableUnsafe { observer ->
+                    observer.onSubscribe(
+                        Disposable {
+                            isResourceDisposedBeforeInnerCompletable = resource.isDisposed
+                        }
+                    )
+                }
+            }.test()
+
+        observer.dispose()
+
+        assertFalse(isResourceDisposedBeforeInnerCompletable)
+    }
+
+    @Test
+    fun resource_is_not_disposed_WHEN_eager_is_false_and_downstream_received_onComplete() {
+        val resource = Disposable()
+        val upstream = TestCompletable()
+        var isResourceDisposedBeforeInnerCompletable by AtomicBoolean()
+
+        val downstream = completableUsing(resourceSupplier = { resource }, eager = false, sourceSupplier = { upstream })
+
+        downstream.subscribe(
+            object : DefaultCompletableObserver {
+                override fun onComplete() {
+                    isResourceDisposedBeforeInnerCompletable = resource.isDisposed
+                }
+            }
+        )
+
+        upstream.onComplete()
+
+        assertFalse(isResourceDisposedBeforeInnerCompletable)
+    }
+
+    @Test
+    fun resource_is_not_disposed_WHEN_eager_is_false_and_downstream_received_onError() {
```

Comment:
Weird test name, it is hard to understand what you are testing here. `using` by design should invoke `resourceCleanup` but test name says it should not.

```kotlin
@Test
fun resource_is_not_disposed _BEFORE_DOWNSTREAM_ WHEN_eager_is_false_and_downstream_received_onError()
```

Reply:
There are two modes: `eager=true` and `eager=false`. The former should dispose the resource before terminating the downstream, the latter should dispose the resource after terminating the downstream. So, when the downstream receives either `onComplete` or `onError`, the resource should already be disposed if `eager=true` and should be yet disposed if `eager=false`.

This test says - resource is not disposed WHEN downstream received `onError`. Perhaps a better name would be "resource_is_not_yet_disposed_WHEN_eager_is_false_and_downstream_received_onError`.

Reply:
What do you think about `resource_is_not_disposed_BEFORE_DOWNSTREAM_`?

Reply:
I have updated all tests, now it says `disposes_resource_after_downstream_signalled_onError_WHEN_eager_false_and_upstream_produced_error`.